libdns.go

```diff
@@ -1,394 +1,212 @@
-// calls, for example, to maintain consistency with all other [libdns] packages.
-// Helper functions are available in this package to convert between relative
-// and absolute names.
+// grants records portability across zones. As such, record names are partially
+// qualified, i.e. relative to the zone. For example, a record called “sub” in
+// zone “example.com.” represents a fully-qualified domain name (FQDN) of
+// “sub.example.com.”. Implementations should expect that input records conform
+// to this standard, while also ensuring that output records do; adjustments to
+// record names may need to be made before or after provider API calls, for example,
+// to maintain consistency with all other [libdns] packages. Helper functions are
+// available in this package to convert between relative and absolute names;
+// see [RelativeName] and [AbsoluteName].
 //
 // Although zone names are a required input, [libdns] does not coerce any
 // particular representation of DNS zones; only records. Since zone name and
 // records are separate inputs in [libdns] interfaces, it is up to the caller to
-// pair a zone's name with its records in a way that works for them.
+// maintain the pairing between a zone's name and its records.
 //
 // All interface implementations must be safe for concurrent/parallel use,
 // meaning 1) no data races, and 2) simultaneous method calls must result in
-// either both their expected outcomes or an error.
-//
-// For example, if [libdns.RecordAppender.AppendRecords] is called at the same
-// time and two API requests are made to the provider at the same time, the
-// result of both requests must be visible after they both complete; if the
-// provider does not synchronize the writing of the zone file and one request
-// overwrites the other, then the client implementation must take care to
-// synchronize on behalf of the incompetent provider. This synchronization need
-// not be global; for example: the scope of synchronization might only need to
-// be within the same zone, allowing multiple requests at once as long as all of
-// them are for different zones. (Exact logic depends on the provider.)
-//
-// [RFC 7719]: https://datatracker.ietf.org/doc/html/rfc7719
+// either both their expected outcomes or an error. For example, if
+// [libdns.RecordAppender.AppendRecords] is called simultaneously, and two API
+// requests are made to the provider at the same time, the result of both requests
+// must be visible after they both complete; if the provider does not synchronize
+// the writing of the zone file and one request overwrites the other, then the
+// client implementation must take care to synchronize on behalf of the incompetent
+// provider. This synchronization need not be global; for example: the scope of
+// synchronization might only need to be within the same zone, allowing multiple
+// requests at once as long as all of them are for different zone. (Exact logic
+// depends on the provider.)
+//
```

Comment:
Can we mention providers can optionally implement io.Closer, for when they keep state around that should be cleaned up when done with the provider. Or perhaps a `Close()` method without error? Useful with providers that create an http.Client for which idle connections are kept around. Their `Close` would call `http.Client.CloseIdleConnections`.

Having

Reply:
Somewhat related to the batching discussions in #154.